pkg/fleetautoscalers/fleetautoscalers.go

```diff
@@ -224,3 +234,298 @@ func applyBufferPolicy(b *autoscalingv1.BufferPolicy, f *agonesv1.Fleet) (int32,
+func scaleDownLimited(f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount, key string, isCounter bool, replicas int32,
+	aggCapacity, maxCapacity int64) (int32, bool, error) {
+	// Game Servers in order of deletion on scale down
+	gameServers, err := getSortedGameServers(f, gameServerLister, nodeCounts)
+	if err != nil {
+		return 0, false, err
+	}
+	for _, gs := range gameServers {
+		if aggCapacity <= maxCapacity {
+			break
+		}
+		switch isCounter {
+		case true:
+			if counter, ok := gs.Status.Counters[key]; ok {
+				aggCapacity -= counter.Capacity
+			}
+		case false:
+			if list, ok := gs.Status.Lists[key]; ok {
+				aggCapacity -= list.Capacity
+			}
+		}
+		replicas--
+	}
+
+	if replicas < 0 { // This shouldn't ever happen, but putting it here just in case.
+		replicas = 0
+	}
+
+	return replicas, true, nil
+}
+
+func scaleLimited(scale int, f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount, key string, isCounter bool, replicas int32,
+	capacity, aggCapacity, minCapacity, maxCapacity int64) (int32, bool, error) {
+
+	switch scale {
+	case 1: // scale up
+		return scaleUpLimited(replicas, capacity, aggCapacity, minCapacity)
+	case -1: // scale down
+		return scaleDownLimited(f, gameServerLister, nodeCounts, key, isCounter, replicas,
+			aggCapacity, maxCapacity)
+	case 0:
+		return replicas, false, nil
+	}
+
+	return 0, false, errors.Errorf("cannot scale due to error in scaleLimited function")
+}
+
+// scaleUp scales up for either Integer or Percentage Buffer.
+func scaleUp(replicas int32, capacity, count, aggCapacity, availableCapacity, maxCapacity,
+	minCapacity, buffer int64) (int32, bool, error) {
+
+	// How much capacity is gained by adding one more replica to the fleet.
+	replicaCapacity := capacity - count
+	if replicaCapacity <= 0 {
+		return 0, false, errors.Errorf("cannot scale up as adding additional replicas does not increase available Capacity")
+	}
+
+	additionalReplicas := int32(math.Ceil((float64(buffer) - float64(availableCapacity)) / float64(replicaCapacity)))
+
+	// Check to make sure we're not limited (over Max Capacity)
+	limited, _ := isLimited(aggCapacity+(int64(additionalReplicas)*capacity), minCapacity, maxCapacity)
+	if limited {
+		additionalReplicas = int32((maxCapacity - aggCapacity) / capacity)
+	}
+
+	return replicas + additionalReplicas, limited, nil
+}
+
+// scaleDown scales down for either Integer or Percentage Buffer.
+func scaleDown(f *agonesv1.Fleet, gameServerLister listeragonesv1.GameServerLister,
+	nodeCounts map[string]gameservers.NodeCount, key string, isCounter bool, replicas int32,
+	availableCapacity, aggCount, aggCapacity, minCapacity, buffer int64) (int32, bool, error) {
+	// Exit early if we're already at MinCapacity or DesiredCapacity to avoid calling
+	// getSortedGameServers if unnecessary
+	if aggCapacity == minCapacity {
```

Comment:
Is this necessary, since the validation is done at line 303 and 343

Reply:
The validation at 303 and 343 that checks if this is below minCapacity, it doesn't check if we're at min Capacity.